### PR TITLE
Limit dependencies to strictly necessary packages.

### PR DESCRIPTION
**Pull request by mkelley**
Make scipy, ads, astroquery, and synphot optional packages, limiting the required packages to numpy and astropy.

* Add new functions and decorators in utils to facilitate testing for optional packages, and raising errors when they are needed.
* Skip tests that require the above packages when they are not available.

To do:
- [x] Verify that tests are working when the above packages are installed (i.e., alldeps).
- [ ] ~Update documentation so that it can be built without optional packages?~

Addresses #382

### CODE REVIEW

File: CHANGES.rst
Comment (mkelley) [inline CHANGES.rst:1]: Change to v0.5.0

File: CHANGES.rst
Comment (mkelley) [inline CHANGES.rst:5]: Add scipy to the list

File: sbpy/activity/gas/productionrate.py
Comment (mkelley) [inline sbpy/activity/gas/productionrate.py:261]: Replace if-branch with `@requires`

File: sbpy/bib/core.py
Comment (mkelley) [inline sbpy/bib/core.py:343]: Fix tests to cover this line.

File: sbpy/utils/tests/test_decorators.py
Comment (mkelley) [inline sbpy/utils/tests/test_decorators.py:10]: Separate function out of `test_requires` and add a second test to cover this line.

File: docs/install.rst
Comment (jianyangli) [inline docs/install.rst:24]: Do you want/need to include a version number for this and some other optional packages?

File: docs/install.rst
Comment (jianyangli) [inline docs/install.rst:28]: Should we just include one package in each entry to keep the format consistent?  Or both `ginga` and `photutils` are needed to interactively enhance images?

File: docs/sbpy/units.rst
Comment (jianyangli) [inline docs/sbpy/units.rst:49]: Do you need this tag before line 62?  I don't know exactly the range of effects for each `..doctest-requires` line, so maybe not?

File: docs/sbpy/units.rst
Comment (jianyangli) [inline docs/sbpy/units.rst:119]: Again, do we need to set up this doctest requires before line 86 and line 107 and line 143?

File: sbpy/activity/dust.py
Comment (jianyangli) [inline sbpy/activity/dust.py:27]: Is this except block necessary?  `scipy` as a variable in this case is not used.

File: docs/install.rst
Comment (mkelley) [reply docs/install.rst:24]: These versions are based on setup.cfg, so if no version number is specified there, I did not list it here.

File: docs/install.rst
Comment (mkelley) [reply docs/install.rst:28]: I can separate these.

File: docs/sbpy/units.rst
Comment (mkelley) [reply docs/sbpy/units.rst:49]: We do.  I fixed this and several other cases.

File: docs/sbpy/units.rst
Comment (mkelley) [reply docs/sbpy/units.rst:119]: Got it.

**Comment by mkelley** [comment]
Cannot build the docs.  Sphinx crashes with the traceback:
```python
Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/sphinx/cmd/build.py", line 298, in build_main
    app.build(args.force_all, args.filenames)
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/sphinx/application.py", line 355, in build
    self.builder.build_update()
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/sphinx/builders/__init__.py", line 293, in build_update
    self.build(to_build,
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/sphinx/builders/__init__.py", line 313, in build
    updated_docnames = set(self.read())
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/sphinx/builders/__init__.py", line 420, in read
    self._read_serial(docnames)
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/sphinx/builders/__init__.py", line 441, in _read_serial
    self.read_doc(docname)
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/sphinx/builders/__init__.py", line 498, in read_doc
    publisher.publish()
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/core.py", line 234, in publish
    self.document = self.reader.read(self.source, self.parser,
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/sphinx/io.py", line 105, in read
    self.parse()
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/readers/__init__.py", line 76, in parse
    self.parser.parse(self.input, document)
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/sphinx/parsers.py", line 81, in parse
    self.statemachine.run(inputlines, document, inliner=self.inliner)
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/parsers/rst/states.py", line 169, in run
    results = StateMachineWS.run(self, input_lines, input_offset,
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/statemachine.py", line 233, in run
    context, next_state, result = self.check_line(
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/statemachine.py", line 445, in check_line
    return method(match, context, next_state)
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/parsers/rst/states.py", line 2785, in underline
    self.section(title, source, style, lineno - 1, messages)
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/parsers/rst/states.py", line 325, in section
    self.new_subsection(title, lineno, messages)
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/parsers/rst/states.py", line 391, in new_subsection
    newabsoffset = self.nested_parse(
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/parsers/rst/states.py", line 279, in nested_parse
    state_machine.run(block, input_offset, memo=self.memo,
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/parsers/rst/states.py", line 195, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/statemachine.py", line 233, in run
    context, next_state, result = self.check_line(
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/statemachine.py", line 445, in check_line
    return method(match, context, next_state)
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/parsers/rst/states.py", line 2785, in underline
    self.section(title, source, style, lineno - 1, messages)
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/parsers/rst/states.py", line 325, in section
    self.new_subsection(title, lineno, messages)
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/parsers/rst/states.py", line 391, in new_subsection
    newabsoffset = self.nested_parse(
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/parsers/rst/states.py", line 279, in nested_parse
    state_machine.run(block, input_offset, memo=self.memo,
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/parsers/rst/states.py", line 195, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/statemachine.py", line 233, in run
    context, next_state, result = self.check_line(
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/statemachine.py", line 445, in check_line
    return method(match, context, next_state)
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/parsers/rst/states.py", line 2355, in explicit_markup
    nodelist, blank_finish = self.explicit_construct(match)
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/parsers/rst/states.py", line 2367, in explicit_construct
    return method(self, expmatch)
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/parsers/rst/states.py", line 2104, in directive
    return self.run_directive(
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/parsers/rst/states.py", line 2154, in run_directive
    result = directive_instance.run()
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/pytest_doctestplus/sphinx/doctestplus.py", line 23, in run
    if re.match('win32', self.content[0]):
  File "/home/docs/checkouts/readthedocs.org/user_builds/sbpy/envs/383/lib/python3.10/site-packages/docutils/statemachine.py", line 1136, in __getitem__
    return self.data[i]
IndexError: list index out of range
```
which doesn't mean anything to me.  I can build ffa51150, so it is definitely something I changed with sbpy.  I will look into this some more.  Meanwhile, the rest of the checks are good.

File: docs/sbpy/units.rst
Comment (mkelley) [reply docs/sbpy/units.rst:49]: The directive is needed for every code-block that is affected by the presence of whatever packages are given.  There are ways to skip all doctests in a Python source file, but I don't see a way to do it for a whole documentation source file, so I have been marking block-by-block.

File: sbpy/activity/gas/productionrate.py
Comment (mkelley) [inline sbpy/activity/gas/productionrate.py:262]: Fix: The second string is a message.

File: docs/sbpy/spectroscopy/index.rst
Comment (jianyangli) [inline docs/sbpy/spectroscopy/index.rst:47]: What's the purpose of lines 33 - 37?  The three lines 35 - 37 are identical as lines 41 - 43.

Comment (mkelley) [reply docs/sbpy/spectroscopy/index.rst:47]: These imports are needed for astropy versions < 5.3 because the next doctest block only executes for versions >=5.3.  I'll add a comment to this effect to the documentation file.

**Comment by mkelley** [comment]
Resolving the documentation build error is pending some more information on the problem via  scientific-python/pytest-doctestplus#219

File: docs/sbpy/units.rst
Comment (mkelley) [reply docs/sbpy/units.rst:49]: Thanks to the pytest-doctestplus maintainers, I have learned that whole documentation files can be marked with a package requirement in setup.cfg via:
```
doctest_subpackage_requires =
    docs/sbpy/photometry.rst = astroquery
```

File: sbpy/activity/gas/core.py
Comment (jianyangli) [inline sbpy/activity/gas/core.py:30]: Do you still need to define variable `scipy` and `PPoly` if scipy is not imported, if you use `requires` decorator to test the import of scipy?

Of course there is no problem to define them, just curious.

File: sbpy/activity/gas/core.py
Comment (jianyangli) [inline sbpy/activity/gas/core.py:620]: Is `@requires("scipy")` missing here?

File: sbpy/activity/gas/core.py
Comment (jianyangli) [inline sbpy/activity/gas/core.py:624]: `@requires("scipy") missing here?

File: sbpy/activity/gas/tests/test_core.py
Comment (jianyangli) [inline sbpy/activity/gas/tests/test_core.py:110]: I didn't closely follow the development of the gas module.  But looks like there is no `.volume_density` method in `Haser`.  Is that why this test is removed here?  But it was not removed in this PR, maybe in the previous PR but this test was left here?

File: sbpy/activity/gas/tests/test_core.py
Comment (jianyangli) [inline sbpy/activity/gas/tests/test_core.py:85]: Just to confirm: Is this and other `importorskip` to test various package set up?

File: sbpy/calib/tests/test_sun.py
Comment (jianyangli) [inline sbpy/calib/tests/test_sun.py:160]: The above codecov warning is very strange.

File: sbpy/data/tests/test_orbit_remote.py
Comment (jianyangli) [inline sbpy/data/tests/test_orbit_remote.py:116]: You have importorskip before all the tests, do you still need one here?

File: sbpy/photometry/core.py
Comment (jianyangli) [inline sbpy/photometry/core.py:15]: I don't think `.from_obs` method uses `scipy`.  Also, `LinearPhaseFunc._phase_integral` is inherited from `DiskIntegratedPhaseFunc`.  Does it need to be included here?

File: sbpy/photometry/iau.py
Comment (jianyangli) [inline sbpy/photometry/iau.py:11]: I see you moved the `__all__` line to the beginning of this file.  In many other files, it's after the imports.  This is only a style issue and doesn't really matter, but should we keep a consistency internally?

File: sbpy/photometry/tests/test_bandpass.py
Comment (jianyangli) [inline sbpy/photometry/tests/test_bandpass.py:50]: Will `bandpass` still raise this error after using `@requires` decorator?

File: sbpy/spectroscopy/sources.py
Comment (jianyangli) [inline sbpy/spectroscopy/sources.py:1]: Should lines 278 and 338 be decorated by `@requires('synphot')` as well?

File: sbpy/units/tests/test_core.py
Comment (jianyangli) [inline sbpy/units/tests/test_core.py:1]: Have a general question about how to set up the test in the case when the functions being called inside a function depend on an optional package.

For example, function `a` doesn't directly uses `synphot`, but `a` calls function `b`, and `b` uses `synphot`.  Should we leave everything (check the availability of a module, test) about optional packages to the functions that directly use the module, and don't do this check for higher level functions or classes?

File: sbpy/utils/tests/__init__.py
Comment (jianyangli) [inline sbpy/utils/tests/__init__.py:1]: Do you really need this empty file here?  Or you'd rather keep it as a placeholder for the future?

File: setup.cfg
Comment (jianyangli) [inline setup.cfg:99]: Not sure how I should review this block.

1. Why do we only specify astroquery here, not others such as scipy?
2. Do you want me to go through all the document files to see if there is anything missing here?

File: sbpy/calib/tests/test_core.py
Comment (jianyangli) [inline sbpy/calib/tests/test_core.py:292]: This line can be removed

File: docs/sbpy/data/orbit.rst
Comment (jianyangli) [inline docs/sbpy/data/orbit.rst:166]: Should this remote-data derivative be removed if the test is now using hardcoded orbital elements?  Same for line 194.

File: docs/sbpy/units.rst
Comment (jianyangli) [reply docs/sbpy/units.rst:49]: Thanks.  I guess we don't have to adopt to this way and clean up all relevant files now, if we want to merge this PR as soon as possible.

Comment (mkelley) [reply docs/sbpy/units.rst:49]: Yeah, it is just an option we could take.

File: sbpy/activity/gas/tests/test_core.py
Comment (mkelley) [reply sbpy/activity/gas/tests/test_core.py:110]: I think I removed it because volume_density is required for Haser's column_density tests.  I'm not sure now if that was the right approach or not.

File: sbpy/activity/gas/tests/test_core.py
Comment (mkelley) [reply sbpy/activity/gas/tests/test_core.py:85]: Yes, if scipy is not available, the test is skipped.

File: sbpy/data/tests/test_orbit_remote.py
Comment (mkelley) [reply sbpy/data/tests/test_orbit_remote.py:116]: The only importorskip I see before all the tests is for astroquery, this one is just or pyoorb.

File: sbpy/photometry/core.py
Comment (mkelley) [reply sbpy/photometry/core.py:15]: This skips the doctest in from_obs's docstring, which does need scipy for the model fitting.

File: sbpy/photometry/iau.py
Comment (mkelley) [reply sbpy/photometry/iau.py:11]: PEP8 recommends "dunders" like __all__ before imports: https://peps.python.org/pep-0008/#module-level-dunder-names   We generally should try to follow PEP8, but it's mostly just helpful, not required.

File: sbpy/photometry/tests/test_bandpass.py
Comment (mkelley) [reply sbpy/photometry/tests/test_bandpass.py:50]: Right, and this test is just verifying that.

File: sbpy/spectroscopy/sources.py
Comment (mkelley) [reply sbpy/spectroscopy/sources.py:1]: They could be, but because the __init__ and from_* methods are decorated with it, no execution can get to this point without synphot.

File: sbpy/units/tests/test_core.py
Comment (mkelley) [reply sbpy/units/tests/test_core.py:1]: I guess just consider the user experience for each?  But also consider what happens if function b is later revised and no longer needs synphot.  I guess then the requirement on a will be obsolete and might be difficult  to find.  So, best to leave it to the point the package is actually used?

File: setup.cfg
Comment (mkelley) [reply setup.cfg:99]: This is specifically for the work-around documented in the comments.  If you wanted to verify that this is needed, then comment out any of lines 94 to 98 and run the tests.

File: sbpy/activity/gas/core.py
Comment (mkelley) [reply sbpy/activity/gas/core.py:30]: PPoly is used in the type annotations.  I can probably remove scipy.

File: sbpy/calib/tests/test_core.py
Comment (mkelley) [reply sbpy/calib/tests/test_core.py:292]: done

File: docs/sbpy/data/orbit.rst
Comment (mkelley) [reply docs/sbpy/data/orbit.rst:166]: I'd like to keep them here as a reminder until the doctest-requires issue is addressed.

File: sbpy/calib/tests/test_sun.py
Comment (mkelley) [reply sbpy/calib/tests/test_sun.py:160]: It is because remote tests are not run as part of our PR CI testing.

File: sbpy/units/tests/test_core.py
Comment (jianyangli) [reply sbpy/units/tests/test_core.py:1]: I agree.

**Review by jianyangli — APPROVED** [review]
I don't see any obvious problems or questions now.  I think it's ready to deploy.

**Comment by mkelley** [comment]
Thanks, @jianyangli !

**Pull request merged**
